src/main/java/com/ldtteam/structurize/client/gui/AbstractBlueprintManipulationWindow.java

```diff
@@ -0,0 +1,365 @@
+import static com.ldtteam.structurize.api.util.constant.WindowConstants.*;
+
+/**
+ * BuildTool window.
+ */
+public abstract class AbstractBlueprintManipulationWindow extends AbstractWindowSkeleton
+{
+    /**
+     * The id of the blueprint type in here.
+     */
+    private final String bluePrintId;
+
+    /**
+     * Settings scrolling list.
+     */
+    protected ScrollingList settingsList;
+
+    /**
+     * Ground style of the caller.
+     */
+    protected int groundstyle;
+
+    /**
+     * How long the UI is open.
+     */
+    private int openTicks = 0;
+
+    /**
+     * Creates a window build tool.
+     * This requires X, Y and Z coordinates.
+     * If a structure is active, recalculates the X Y Z with offset.
+     * Otherwise the given parameters are used.
+     *
+     * @param pos coordinate.
+     * @param groundstyle one of the GROUNDSTYLE_ values.
+     */
+    public AbstractBlueprintManipulationWindow(@NotNull final String resourceId, @Nullable final BlockPos pos, final int groundstyle, final String blueprintId)
+    {
+        super(resourceId);
+        this.groundstyle = groundstyle;
+        this.bluePrintId = blueprintId;
+
+        if (pos != null && RenderingCache.getOrCreateBlueprintPreviewData(blueprintId).pos == null)
+        {
+            RenderingCache.getOrCreateBlueprintPreviewData(blueprintId).pos = pos;
+            adjustToGroundOffset();
+        }
+
+        // Register all necessary buttons with the window.
+        registerButton(BUTTON_CONFIRM, this::confirmClicked);
+        registerButton(BUTTON_CANCEL, this::cancelClicked);
+        registerButton(BUTTON_LEFT, this::moveLeftClicked);
+        registerButton(BUTTON_MIRROR, this::mirror);
+        registerButton(BUTTON_RIGHT, this::moveRightClicked);
+        registerButton(BUTTON_BACKWARD, this::moveBackClicked);
+        registerButton(BUTTON_FORWARD, this::moveForwardClicked);
+        registerButton(BUTTON_UP, this::moveUpClicked);
+        registerButton(BUTTON_DOWN, this::moveDownClicked);
+        registerButton(BUTTON_ROTATE_RIGHT, this::rotateRightClicked);
+        registerButton(BUTTON_ROTATE_LEFT, this::rotateLeftClicked);
+        registerButton(BUTTON_SETTINGS, this::settingsClicked);
+
+        settingsList = findPaneOfTypeByID("settinglist", ScrollingList.class);
+        updateRotationState();
+
+        findPaneOfTypeByID("tip", Text.class).setVisible(pos != null);
+    }
+
+
+    /**
+     * On clicking the red cancel button.
+     */
+    protected abstract void cancelClicked();
+
+    /**
+     * On clicking confirm for placement.
+     */
+    protected abstract void confirmClicked();
+
+    @Override
+    public void onUpdate()
+    {
+        super.onUpdate();
+        if (openTicks++ >= 20 * 10)
+        {
+            findPaneOfTypeByID("tip", Text.class).setVisible(false);
+        }
+    }
+
+    @Override
+    public boolean onKeyTyped(final char ch, final int key)
+    {
+        if (ch == '\u0000')
```

Comment:
Handle implementation of the mirrot button here somewhere as well.

Reply:
Move this logic to an option in BlockUI. Possibly somehow linked to keybinds in the key binding menu? Not sure how that works.